### PR TITLE
TestFramework: Support issue ID assertion

### LINKED ISSUE #8623: TestFramework: Support issue ID assertion
Re-enable asserting issue ID between primary and secondary locations:

```
        public void UnexpectedSecondaryIssueWrongId() =>
            VerifyThrows<AssertFailedException>("""
                public class UnexpectedSecondary
                {
                    public void Test(bool a, bool b)
                    {
                        // Secondary@+1 [myWrongId]
                        if (a == a) // Noncompliant [myId]
                        { }
                    }
                }
                """,
                "CSharp*: Unexpected secondary issue [myId] on line 7, span (6,12)-(6,13) with message ''." + Environment.NewLine +
                "See output to see all actual diagnostics raised on the file");
```

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #8623

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs [hunk 1]
```diff
@@ -186,7 +189,9 @@ private static void VerifyBuildErrors(ImmutableArray<Diagnostic> diagnostics, Co
         private static IEnumerable<IssueLocationPair> MatchPairs(CompilationIssues actual, CompilationIssues expected)
         {
             var ret = new List<IssueLocationPair>();
-            foreach (var key in actual.UniqueKeys())
+            // Process project-level issues last to properly match path-based issues first.
+            // Then process primary issues first, so we can update their IssueId for the purpose of matching secondary issues with their IssueId
```
Comment (gregory-paidis-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs:193]: I would write this comment in the order of processing:

```csharp
// First, process primary issues to update their IssueId for secondary-matching
// Second, process secondary issues
// Last, process project-level issues that ignore path-based matching

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs [hunk 2]
```diff
@@ -198,15 +203,26 @@ private static IEnumerable<IssueLocationPair> MatchDifferences(List<IssueLocatio
...
+                    .OrderBy(x => actual.IssueId == x.IssueId ? 0 : 1)
+                    .ThenBy(x => Math.Abs(actual.Start ?? 0 - x.Start ?? 0))
+                    .ThenBy(x => Math.Abs(actual.Length ?? 0 - x.Length ?? 0))
+                    .FirstOrDefault();
+                if (expected is not null)
+                {
+                    actual.UpdatePrimaryIssueIdFrom(expected);
+                    expectedIssues.Remove(expected);
+                }
                 yield return new(actual, expected); // expected can be null
```
Comment (gregory-paidis-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs:226]: Do you think we need this comment?
Right above you check nullability, and afterwards you unconditionally yield.
So it seems obvious that both null/not-null are valid.

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/IssueValidation/IssueLocation.cs [hunk 3]
```diff
@@ -52,17 +56,25 @@ private IssueLocation(string message, Location location)
...
-        && issue.FilePath == FilePath
+        && (issue.FilePath == string.Empty || FilePath == string.Empty || issue.FilePath == FilePath)
         && issue.LineNumber == LineNumber
         && issue.IsPrimary == IsPrimary
+        && (IsPrimary || issue.IssueId == IssueId)   // We ignore issueId for primary issues, as we need to match them only for secondary
         && (issue.RuleId is null || RuleId is null || issue.RuleId == RuleId)
         && (issue.Message is null || Message is null || issue.Message == Message)
-        && (issue.IssueId is null || IssueId is null || issue.IssueId == IssueId)   // For consistency, Roslyn issues will not have IDs
         && (issue.Start is null || Start is null || issue.Start == Start)
         && (issue.Length is null || Length is null || issue.Length == Length);
```
Comment (gregory-paidis-sonarsource) [inline analyzers/tests/SonarAnalyzer.TestFramework/Verification/IssueValidation/IssueLocation.cs:79]: I think we should add two helper methods here to make the big chain more readable:

```csharp
        private bool EqualOrNull(string first, string second) =>
            first is null || second is null || first == second;

        private bool EqualOrEmpty(string first, string second) =>
            first == string.Empty || second == string.Empty || first == second;
```

**Review by gregory-paidis-sonarsource — APPROVED** [review]
LGTM, I left a couple of suggestions, please take a look

File: analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.TestFramework/Verification/DiagnosticVerifier.cs:193]: I kept it on two lines, and reworded

**Pull request merged**